### PR TITLE
Topic recognition-Siamese network classifier

### PR DESCRIPTION
**Pull request by aniketgupta17**
# Pull Request Description

## Purpose
This pull request implements a Siamese Network that is  trained to generate embeddings of images of the ADNI dataset. A classifier (fully connected layers) is added at the end of the trained Siamese Network to classify Alzheimer's Disease on the dataset.

### Description
Task Chosen: 7. Create a classifier based on a Siamese network [10] to classify either Alzheimer’s disease (normal and AD) of the ADNI brain data set having an accuracy of around 0.8 on the test set [Hard Difficulty].

Model: The backbone architecture of the Siamese network is based on PyTorch's ResNet-18 with modifications. The classifier's architecture consists of fully connected layers to do classification tasks. A trained Siamese model sits at the end of the classifier with its weights frozen.

### Dataset Preparation (dataset.py):
This code defines a custom dataset class, SiameseADNIDataset for loading and processing ADNI dataset images for use in Siamese Network training and testing. It also provides functions to get train and test datasets from a specified data path. and  validating. This also include custom dataloader to handle Pair Dataset for Siamese model.

### Network classifier Modules (modules.py):
This code defines a Siamese Network architecture for image similarity comparison. The network is built upon the ResNet-18 architecture with modifications to handle grayscale images. It includes custom weight initialization and forward pass methods.

### Training Script (train.py):
This script is used to train a Siamese Network model on a dataset, with support for validation .It includes training and validation loops, model saving, and TensorBoard logging.

### Testing  Script (predict.py):
This script is used to test a Siamese Network model on a dataset. It loads a trained Siamese Network model, evaluates its performance on a test dataset, and reports the accuracy and loss.

Kindly assess the alterations made in this pull request, and if you find them satisfactory, contemplate merging the request. In the event that you have particular adjustments or enhancements in mind, I would appreciate your valuable feedback.

### CODE REVIEW

**Comment by nathasha-naranpanawa** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme: Good
- clear flow of information
- algorithm and training choices explained well
- loss and accuracy plots provided for the siamese, but only for the training set, not very useful

Commit messages: Good

Code: 
- design is OK
- well commented
- only implemented a siamese network, no classifier

Functionality/Performance: 
- **only performing similarity detection between pairs of images, no classification of single images**
- achieves good accuracy for similarity detection with the siamese

General comments:
- **algorithm solves the respective problem only half-way**
- you need to have added a classifier that can classify an individual image as AD or NC with their features extracted by the trained siamese network

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem partially no classification of single images -3
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good but missing usage -1
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
